### PR TITLE
[WIP] Search pootle path using regex

### PR DESCRIPTION
**Pull request by phlax**
At the moment when using Unit.objects.get_by_path it uses extra fields. These fields cause a headache with the queryset as store__pootle_path drops out of the scope of the qs.

Also, if pootle_path is not set it currently adds a search of like "/%" which is uncessary.

This PR:
- adds a PathResolver using django's url resolver
- adds regex filtering of qs rather than using extra fields
- removes unnecessary pootle_path filtering 
- adds some tests for above

**just working on some of the code around split_pootle_path**

### CODE REVIEW

File: pootle/core/url_helpers.py [hunk 1]
```diff
@@ -12,7 +12,83 @@
...
+            return "/%s" % self._path.strip("/")
+        return "/%s/" % self._path.strip("/")
+
+    @cached_property
+    def kwa(self):
+        return resolve(self.path).kwargs
+
+    @property
+    def directory(self):
+        if not self.path:
```
Comment (unho) [inline pootle/core/url_helpers.py]: Shouldn't this be `self._path`?

Comment (phlax) [inline pootle/core/url_helpers.py]: no, i dont thinks so - if self.path is empty dont try to parse

File: pootle/core/url_helpers.py [hunk 2]
```diff
@@ -12,7 +12,83 @@
...
+        return self.kwa.get("language_code", None) or None
+
+    @property
+    def project(self):
+        if not self.path:
+            return None
+        return self.kwa.get("project_code", None) or None
+
+    @property
+    def regex(self):
```
Comment (unho) [inline pootle/core/url_helpers.py]: Adding some docstring here would be helpful as it is not obvious what is this meant for.

File: pootle/core/url_helpers.py [hunk 3]
```diff
@@ -12,7 +12,83 @@
...
 import urlparse
 
-from django.core.urlresolvers import reverse
+from django.core.urlresolvers import reverse, resolve
+from django.utils.functional import cached_property
+
+
+class PathResolver(object):
+    """Converts a request.path into parts for matching pootle_path
+    with a regex
```
Comment (unho) [inline pootle/core/url_helpers.py]: I would rephrase this:

Normalizes the request.path, extracts its parts and provides a regex useful to match the path.

Or use a even more explanatory docstring with examples.

File: pootle/apps/pootle_store/models.py [hunk 4]
```diff
@@ -40,7 +40,8 @@
 from pootle.core.models import Revision
 from pootle.core.storage import PootleFileSystemStorage
 from pootle.core.search import SearchBroker
-from pootle.core.url_helpers import get_editor_filter, split_pootle_path
+from pootle.core.url_helpers import (
+    get_editor_filter, split_pootle_path, PathResolver)
```
Comment (unho) [inline pootle/apps/pootle_store/models.py]: Nit: Use styleguide recommendations on how to wrap long lines: "If you break in a statement with parentheses or braces, align to the braces" in http://docs.translatehouse.org/projects/translate-toolkit/en/latest/developers/styleguide.html#continuing-long-statements

File: pootle/core/url_helpers.py (hunk 2 above)
Comment (unho) [inline pootle/core/url_helpers.py]: Nit: Instead of adding the regex capability on the commit where `PathResolver` is added, I would add it on the "Use PathResolver.regex when filtering by pootle_path" commit. This way it is easier to see why some feature has been initially added.

File: pootle_pytest/fixtures/models/unit.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
+
```
Comment (unho) [inline pootle_pytest/fixtures/models/unit.py]: No license notice?

File: pootle_pytest/plugin.py [hunk 6]
```diff
@@ -46,7 +48,8 @@
 PARAMETERS = (
     ("update_store_test_names", UPDATE_STORE_TESTS),
     ("file_import_failure_names", FILE_IMPORT_FAIL_TESTS),
-    ("wordcount_names", WORDCOUNT_TESTS))
+    ("wordcount_names", WORDCOUNT_TESTS),
+    ("unit_path_resolver_names", UNIT_PATH_RESOLVER_TESTS))
```
Comment (unho) [inline pootle_pytest/plugin.py]: Nit: We should go back to having the opening and closing brackets on its own line to avoid the diff to include stupid changes like this.

File: tests/models/unit.py [hunk 7]
```diff
@@ -281,3 +281,41 @@ def test_accept_suggestion_update_wordcount(it_tutorial_po, system):
...
+            assert parsed is None
+        else:
+            assert parsed == ""
+
+    units = Unit.objects.order_by("pk")
+    if resolved.regex:
+        expected = units.filter(store__pootle_path__regex=resolved.regex)
+    else:
+        expected = units.all()
+    if result.get("project"):
```
Comment (unho) [inline tests/models/unit.py]: I would add a blank line before this `if` for readability.

File: pootle/core/url_helpers.py (hunk 1 above)
Comment (unho) [inline pootle/core/url_helpers.py]: My fault, I always forget to check if there is any attribute with this name.

File: pootle/core/decorators.py [hunk 8]
```diff
@@ -24,10 +24,10 @@
 from pootle_project.models import Project, ProjectSet, ProjectResource
 from pootle_store.models import Store
 from pootle_translationproject.models import TranslationProject
-from virtualfolder.helpers import extract_vfolder_from_path
+from virtualfolder.helpers import extract_vfolder_from_request
 
 from .exceptions import Http400
-from .url_helpers import split_pootle_path
```
Comment (unho) [inline pootle/core/decorators.py]: There is another usage of `split_pootle_path` in this file, so either get the import back, or replace it.

**Comment by phlax** [comment]
This is no longer required
